File: backend/src/main/java/com/example/knowledgehub/backend/resource/LearningResourceController.java
package com.example.knowledgehub.backend.resource;

import com.example.knowledgehub.backend.category.Category;
import com.example.knowledgehub.backend.category.CategoryRepository;
import com.example.knowledgehub.backend.common.NotFoundException;
import com.example.knowledgehub.backend.resource.LearningResourceDtos.LearningResourceCreateRequest;
import com.example.knowledgehub.backend.resource.LearningResourceDtos.LearningResourceResponse;
import com.example.knowledgehub.backend.resource.LearningResourceDtos.LearningResourceUpdateRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/resources")
@CrossOrigin(origins = "*")
public class LearningResourceController {

    private final LearningResourceRepository resourceRepository;
    private final CategoryRepository categoryRepository;

    public LearningResourceController(LearningResourceRepository resourceRepository, CategoryRepository categoryRepository) {
        this.resourceRepository = resourceRepository;
        this.categoryRepository = categoryRepository;
    }

    @GetMapping
    public List<LearningResourceResponse> getAll() {
        return resourceRepository.findAll().stream()
                .map(LearningResourceDtos::toResponse)
                .collect(Collectors.toList());
    }

    @GetMapping("/{id}")
    public LearningResourceResponse getOne(@PathVariable Long id) {
        LearningResource resource = resourceRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Resource " + id + " not found"));
        return LearningResourceDtos.toResponse(resource);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public LearningResourceResponse create(@Valid @RequestBody LearningResourceCreateRequest request) {
        Category category = null;
        if (request.categoryId() != null) {
            category = categoryRepository.findById(request.categoryId())
                    .orElseThrow(() -> new NotFoundException("Category " + request.categoryId() + " not found"));
        }
        LearningResource entity = LearningResourceDtos.fromCreateRequest(request, category);
        LearningResource saved = resourceRepository.save(entity);
        return LearningResourceDtos.toResponse(saved);
    }

    @PutMapping("/{id}")
    public LearningResourceResponse replace(@PathVariable Long id, @Valid @RequestBody LearningResourceCreateRequest request) {
        LearningResource existing = resourceRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Resource " + id + " not found"));
        Category category = null;
        if (request.categoryId() != null) {
            category = categoryRepository.findById(request.categoryId())
                    .orElseThrow(() -> new NotFoundException("Category " + request.categoryId() + " not found"));
        }
        existing.setName(request.name());
        existing.setDescription(request.description());
        existing.setUrl(request.url());
        existing.setCategory(category);
        LearningResource saved = resourceRepository.save(existing);
        return LearningResourceDtos.toResponse(saved);
    }

    @PatchMapping("/{id}")
    public LearningResourceResponse update(@PathVariable Long id, @RequestBody LearningResourceUpdateRequest request) {
        LearningResource existing = resourceRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Resource " + id + " not found"));
        if (request.name() != null) {
            existing.setName(request.name());
        }
        if (request.description() != null) {
            existing.setDescription(request.description());
        }
        if (request.url() != null) {
            existing.setUrl(request.url());
        }
        if (request.categoryId() != null) {
            Category category = categoryRepository.findById(request.categoryId())
                    .orElseThrow(() -> new NotFoundException("Category " + request.categoryId() + " not found"));
            existing.setCategory(category);
        }
        LearningResource saved = resourceRepository.save(existing);
        return LearningResourceDtos.toResponse(saved);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable Long id) {
        if (!resourceRepository.existsById(id)) {
            throw new NotFoundException("Resource " + id + " not found");
        }
        resourceRepository.deleteById(id);
    }
}

File: backend/src/main/java/com/example/knowledgehub/backend/resource/LearningResourceController.kt
package com.example.knowledgehub.backend.resource

import com.example.knowledgehub.backend.category.CategoryRepository
import com.example.knowledgehub.backend.common.NotFoundException
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/resources")
@CrossOrigin(origins = ["*"])
class LearningResourceController(
    private val resourceRepository: LearningResourceRepository,
    private val categoryRepository: CategoryRepository,
) {

    @GetMapping
    fun getAll(
        @RequestParam(required = false) categoryId: Long?,
        @RequestParam(required = false) categoryCode: String?,
    ): List<LearningResourceResponse> {
        val resources = when {
            categoryId != null -> {
                val cat = categoryRepository.findById(categoryId)
                    .orElseThrow { NotFoundException("Category $categoryId not found") }
                resourceRepository.findAllByCategory(cat)
            }
            !categoryCode.isNullOrBlank() -> resourceRepository.findAllByCategoryCode(categoryCode)
            else -> resourceRepository.findAll()
        }
        return resources.map { it.toResponse() }
    }

    @GetMapping("/{id}")
    fun getOne(@PathVariable id: Long): LearningResourceResponse =
        resourceRepository.findById(id).orElseThrow { NotFoundException("Resource $id not found") }.toResponse()

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(@Valid @RequestBody request: LearningResourceCreateRequest): LearningResourceResponse {
        val category = categoryRepository.findById(request.categoryId)
            .orElseThrow { NotFoundException("Category ${'$'}{request.categoryId} not found") }
        val entity = LearningResource(
            category = category,
            title = request.title,
            shortDescription = request.shortDescription,
            resourceUrl = request.resourceUrl,
            iconUrl = request.iconUrl,
            orderIndex = request.orderIndex,
        )
        return resourceRepository.save(entity).toResponse()
    }

    @PutMapping("/{id}")
    fun replace(@PathVariable id: Long, @Valid @RequestBody request: LearningResourceCreateRequest): LearningResourceResponse {
        val existing = resourceRepository.findById(id).orElseThrow { NotFoundException("Resource $id not found") }
        val category = categoryRepository.findById(request.categoryId)
            .orElseThrow { NotFoundException("Category ${'$'}{request.categoryId} not found") }
        existing.category = category
        existing.title = request.title
        existing.shortDescription = request.shortDescription
        existing.resourceUrl = request.resourceUrl
        existing.iconUrl = request.iconUrl
        existing.orderIndex = request.orderIndex
        return resourceRepository.save(existing).toResponse()
    }

    @PatchMapping("/{id}")
    fun update(@PathVariable id: Long, @RequestBody request: LearningResourceUpdateRequest): LearningResourceResponse {
        val existing = resourceRepository.findById(id).orElseThrow { NotFoundException("Resource $id not found") }
        if (request.categoryId != null) {
            val category = categoryRepository.findById(request.categoryId)
                .orElseThrow { NotFoundException("Category ${'$'}{request.categoryId} not found") }
            existing.category = category
        }
        if (request.title != null) existing.title = request.title
        if (request.shortDescription != null) existing.shortDescription = request.shortDescription
        if (request.resourceUrl != null) existing.resourceUrl = request.resourceUrl
        if (request.iconUrl != null) existing.iconUrl = request.iconUrl
        if (request.orderIndex != null) existing.orderIndex = request.orderIndex
        return resourceRepository.save(existing).toResponse()
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: Long) {
        if (!resourceRepository.existsById(id)) throw NotFoundException("Resource $id not found")
        resourceRepository.deleteById(id)
    }
}
